src/main/java/rx/observables/SyncOnSubscribe.java

```diff
@@ -0,0 +1,499 @@
+ */
+
+package rx.observables;
+
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.concurrent.atomic.AtomicLong;
+
+import rx.Observable.OnSubscribe;
+import rx.Producer;
+import rx.Subscriber;
+import rx.Subscription;
+import rx.annotations.Experimental;
+import rx.functions.Action0;
+import rx.functions.Action1;
+import rx.functions.Action2;
+import rx.functions.Func0;
+import rx.functions.Func2;
+import rx.internal.operators.BackpressureUtils;
+
+/**
+ * A utility class to create OnSubscribe functions which manages the back pressure concerns that
+ * otherwise aren't handled by {@link rx.Observable#create(OnSubscribe)
+ * Observable.create(OnSubscribe)}. Note that the behavior in the {@code next} argument must
+ * serialize all calls to the downstream subscriber.
+ *
+ * @param <S>
+ *            the type of the user-define state used in {@link #generateState() generateState(S)} ,
+ *            {@link #next(Object, Subscriber) next(S, Subscriber)}, and
+ *            {@link #onUnsubscribe(Object) onUnsubscribe(S)}.
+ * @param <T>
+ *            the type {@code Subscribers} that will be compatible with {@code this}.
+ */
+@Experimental
+public abstract class SyncOnSubscribe<S, T> implements OnSubscribe<T> {
+    /**
+     * Used to signify when there is unsubscribe behavior
+     */
+    protected Action1<? super S> onUnsubscribe;
+
+    /**
+     * Produces an initial state value. The default implementation returns {@code null}.
+     * 
+     * @return the initial state value that will be received by the {@link #next(Object, Subscriber)
+     *         next(S, Subscriber)} function as it's first argument on it's first execution.
+     */
+    protected S generateState() {
+        return null;
+    }
+
+    /**
+     * SyncOnSubscribe iterates over this function to produce values in a way compatible with
+     * back-pressure. Implementations may not call {@link rx.Subscriber#onNext(Object)
+     * subscriber.onNext(T)} concurrently. You may serialize onNext calls over more than one thread,
+     * but concurrent calls to {@code onNext(t)} will result in a runtime exception. Calling
+     * {@link rx.Subscriber#onCompleted() subscriber.onCompleted()} or
+     * {@link rx.Subscriber#onError(Throwable) subscriber.onError(Throwable)} will result in the
+     * downstream subscriber unsubscribing and executing
+     * {@link SyncOnSubscribe#onUnsubscribe(Object) onUnsubscribe(S)}.
+     * 
+     * @param state
+     *            the value produced by {@link #generateState()} on the first execution or the value
+     *            returned from the previous execution of {@code next(Object, Subscriber)}
+     * @param subscriber
+     *            down stream data subscriber to receive onNext, onCompleted, or onError
+     * @return the state value to be received by the next execution of this function
+     */
+    protected abstract S next(S state, Subscriber<? super T> subscriber);
```

Comment:
Have you considered exposing only the `Observer` here? Do you want the `next` to add resources to the subscriber or check for its unsubscribed state?